guile: update to 3.0.9, fix ppc; guile2: old guile goes here

barracuda156 opened the pull request:
#### Description

Long due version update for `guile`, patch for `ppc32`.
Old 2.2.7 moved to a new port, `guile2`.

###### Type(s)

- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
macOS 10.8.5 (`x86_64`)
Xcode 5.1.1

macOS 10.6.8 Server (`ppc`, `x86_64`)
Xcode 3.2.6

macOS 10A190 (`ppc`)
Xcode 3.2

Installation of both new `guile` and `guile2` (old 2.2.7) tested on 10.6.8 Intel. Won’t bother repeat on PPC, takes forever.
New `guile` builds on PPC, at least ppc32.

barracuda156 commented:
@kencu The port has no maintainers, so I guess we have to decide anyway. `guile` is stuck on the very old version (and on PPC it didn’t even build as-is).

barracuda156 commented:
Looks like there is a problem with parsing the ending portion of the portfile where a part is commented out inside brackets. I will fix that today.

barracuda156 commented:
@kencu @catap Alternatively, we may either move an old `guile` to `guile2` or the new one to `guile3`.

catap commented:
@barracuda156 does any port depends on `guile`? And if so, can it be build with your update?

barracuda156 commented:
> @barracuda156 does any port depends on `guile`? And if so, can it be built with your update?

@catap Not many; with a previous @2.9.9 (which in fact reports itself as @3.0) I have built `gnotime` and `swig3-guile`, with patching config to look for a correct `guile` version.

@kencu On a side note, why does the macos-10.15 bot report this error?
```
find: /tmp/mpbb: No such file or directory
[11](https://github.com/macports/macports-ports/runs/6059903717?check_suite_focus=true#step:9:11)
Error: Process completed with exit code 1.
```
What does that even mean? )

kencu commented:
> What does that even mean? )

You got me -- no clue :>

kencu commented:
> @kencu @catap Alternatively, we may either move an old `guile` to `guile2` or the new one to `guile3`.

That would be great! Usually, the way this works is the current port has the default name, in this case "guile" and an older version left behind to support some port that needs that version still has a versioned name like "guile18" or "guile2" or "guile22" or something like that.

In the case of guile, the last time somebody tried to roll out an older version (guile18) it turned out to be a helluva mess https://github.com/macports/macports-ports/blob/master/lang/guile18/Portfile but perhaps all the new stuff in MacPorts since then will make it easier to make a guile2 port.

barracuda156 commented:
@kencu Okay, another ten hours, and I can confirm that `--enable-shared` works, and in fact is necessary for dependencies to build. I have enabled it in the portfile. `gnotime` has build successfully using `guile` @3.0.8.

barracuda156 commented:
> > @kencu @catap Alternatively, we may either move an old `guile` to `guile2` or the new one to `guile3`.
> 
> That would be great! Usually, the way this works is the current port has the default name, in this case "guile" and an older version left behind to support some port that needs that version still has a versioned name like "guile18" or "guile2" or "guile22" or something like that.
> 
> In the case of guile, the last time somebody tried to roll out an older version (guile18) it turned out to be more complicated than expected https://github.com/macports/macports-ports/blob/master/lang/guile18/Portfile but perhaps all the new stuff in MacPorts since then will make it easier to make a guile2 port.

@kencu I agree that making an old version as `guile2` appears to be a better idea. Technically it is straightforward, we literally copy the existing portfile into a new port. Should I do that? (Can I even commit a new port?)

kencu commented:
> @kencu I agree that making an old version as `guile2` appears to be a better idea. Technically it is straightforward, we literally copy the existing portfile into a new port. Should I do that? (Can I even commit a new port?)

It's a bit more complicated. The new guile2 port has to be 
1. installed somewhere other than the default place guile is installed 
2. not collide with the guile18 or new guile port 
3. still has to be able to be found by the ports that need to use it when they try to build.

barracuda156 commented:
> > @kencu I agree that making an old version as `guile2` appears to be a better idea. Technically it is straightforward, we literally copy the existing portfile into a new port. Should I do that? (Can I even commit a new port?)
> 
> It's a bit more complicated. The new guile2 port has to be
> 
> 1. installed somewhere other than the default place guile is installed
> 2. not collide with the guile18 or new guile port
> 3. still has to be able to be found by the ports that need to use it when they try to build.

@kencu Got it. Can someone take care of that? These are nuances specific to Macports internals, which I am not aware of. Or if there is a transparent step-wise instruction, I can do that, of course.
I will also appreciate if someone can test `guile` @3.0.8 on a relevant Intel system. I would imagine building on PPC is more problematic, and on Intel every option gonna work as-is, but better be safe.

kencu commented:
> @kencu Got it. Can someone take care of that? These are nuances specific to Macports internals, which I am not aware of. 

The last time I posted a ticket about guile spin-offs being needed, it sat for two or three years until I closed it finally. So here there might be a need to recruit some help. The macports-dev mailing list is often a good place to request such help.

> Or if there is a transparent step-wise instruction, I can do that, of course. 

No such instructions exist to my knowledge.

> I will also appreciate if someone can test `guile` @3.0.8 on a relevant Intel system.  I would imagine building on PPC is more problematic, and on Intel every option gonna work as-is, but better be safe.

This of course would be a must. Nothing should be pushed to MacPorts tree unless it - at the very least - builds on the three current systems that are actively being supported. Exceptions made for very specific ports like gcc10-bootstrap that are meant to be used on older systems.

barracuda156 commented:
Building `guile` @3.0.8 now on 10.6.8 Rosetta, will leave it to compile overnight, but so far all looks good.

harens commented:
> On a side note, why does the macos-10.15 bot report this error?
> 
> ```
> find: /tmp/mpbb: No such file or directory
> [11](https://github.com/macports/macports-ports/runs/6059903717?check_suite_focus=true#step:9:11)
> Error: Process completed with exit code 1.
> ```
> 
> What does that even mean? )


By default, when a GitHub action fails, any workflows that are still running are cancelled. Running more tests doesn't help much when there are already errors to fix. It's more useful to have the bots running on other PRs.

The macOS 11 build failed, so the 10.15 build was cancelled. Hence the obscure error :)

barracuda156 commented:
`guile` @3.0.8 successfully built on 10.6.8 Rosetta:
```
macmini:~ svacchanda$ port -v installed guile
The following ports are currently installed:
  guile @3.0.8_0 (active) requested_variants='' platform='darwin 10' archs='ppc' date='2022-04-21T02:27:46+0800'
```
Took a while, since in VM.

barracuda156 commented:
@kencu @harens Dependencies fixed, `guile` builds on bots now.

barracuda156 commented:
@mascguy @kencu Do we have reservations with this update? Or should it rather be turned into `guile-devel` or `guile3`? I am perfectly fine with either of the options, as you guys see optimal for Macports.

kencu commented:
I am in charge of nothing, but unless the above comments about making a guile2 port happen are done then many things will break I think.

barracuda156 commented:
> I am in charge of nothing, but unless the above comments about making a guile2 port happen are done then many things will break I think.

What do you think a painless transition could be? Maybe a technically easier, though perhaps less consistent, would be just introduce `guile3` for the time-being, until everything is confirmed to build with it? In such a case no extra PRs required immediately, nothing gonna break.

kencu commented:
It seems to me that it would be just as much trouble to make a non-conflicting guile3 port as it would be to make a non-conflicting guile2 port, so just make the guile2 port. Don’t just make a guile3 port that conflicts with guile (2), as then somebody just needs to come after you and fix it right. But that’s just me.

barracuda156 commented:
Both OpenBSD and NetBSD decided to keep all three major versions of Guile as separate ports:
https://github.com/NetBSD/pkgsrc/tree/trunk/lang
https://github.com/openbsd/ports/tree/master/lang

catap commented:
@barracuda156 just an idea split it into commits:
 - rename glue -> glue2
 - updated dependencies to use glue2
 - introduce glue3

barracuda156 commented:
> @barracuda156 just an idea split it into commits:
> * rename guile -> guile2
> * updated dependencies to use guile2
> * introduce guile3

@mascguy @kencu WDYT?

kencu commented:
> @barracuda156 just an idea split it into commits:
> 
> * rename glue -> glue2
> * updated dependencies to use glue2
> * introduce glue3

this was always the recommended plan, but turned out to be harder than it looks when previously tried.

kencu commented:
oh, whether the newest guile is numbered or not is a small thing. Parallel installs of the various guile versions was the thing thst takes a bit of coding for someone

barracuda156 commented:
> oh, whether the newest guile is numbered or not is a small thing. Parallel installs of the various guile versions was the thing thst takes a bit of coding for someone

What example do we have to avoid conflicts in such cases?

catap commented:
@barracuda156 I suggest to start to figure out: did glue2 and glue3 had compatibily API? If no => separated ports. If yes => you may simple update the port. If ABI is different => you must increase revision of all ports which is depends on it.

barracuda156 commented:
> @barracuda156 I suggest to start to figure out: did glue2 and glue3 had compatibily API? If no => separated ports. If yes => you may simple update the port. If ABI is different => you must increase revision of all ports which is depends on it.

As the very minimum, dependents may need patches to configure scripts to recognize `guile` 3.x (I have done that locally for some).

barracuda156 commented:
Let’s see if it still builds fine, and then I will try amending `guile2` accordingly.

barracuda156 commented:
Okay, so `guile3` builds fine on Intel (CI checks fine, locally 10.6.8 x86_64 built fine), PowerPC bootstrap takes forever, currently running on a PowerMac and in Rosetta, but it built before, so should be good too.

I am gonna adopt the logic from `guile18` to do the same for `guile2`. If that works locally, gonna finalize the PR. (May not be today, but in any case very soon.)

barracuda156 commented:
@kencu Installs fine alongside. Now we need to find out whether dependents should be made to depend on `guile2` or the new one.

```
macmini:bin svacchanda$ sw_vers
ProductName:	Mac OS X Server
ProductVersion:	10.6.8
BuildVersion:	10K549
macmini:bin svacchanda$ ./port -v installed | grep guile
  guile @3.0.8_0 (active) requested_variants='' platform='darwin 10' archs='x86_64' date='2022-11-03T18:40:05+0800'
  guile2 @2.2.7_3 (active) requested_variants='' platform='darwin 10' archs='x86_64' date='2022-11-03T21:20:37+0800'
```

kencu commented:
to test your proposal, you might install only your guile2 port, and no other guiles, and then see if any of the current ports that depend on guile can be made to find it.

Then, install the new guile, and make sure ports configured to use the guile2 port don’t accidentally find and use the new guile instead.

barracuda156 commented:
@kencu What is _desirable,_ to begin with? Would we rather move to new `guile` whatever builds with it – or we want to set all current dependents to use `guile2`?

barracuda156 commented:
@kencu I.e. in one scenario I just need to make sure that all dependents can find `guile2` with or without the new guile installed. In another we need to test what builds with the new one and what fails, and adjust accordingly (this looks more sensible, but it the first choice is a preferred policy, then testing builds gonna be a waste of time).

barracuda156 commented:
`guile` updated to 3.0.9, time to finish this :)

I will move dependencies to use `guile2` for now, so that we can merge this. May take a couple of days, but I wanna finalize this ASAP.

barracuda156 commented:
So as of now, both guile 2.2.7 and guile 3.0.9 confirmed to build fine on CI and locally (x86 and ppc).

I am adding some ports to see how they build, given guiles co-existing.

Notice that some are broken for select systems, so may not build here either, like `gnotime` (broken Big Sur up): https://ports.macports.org/port/gnotime/details
But let us see.

@kencu @ryandesign Could someone help me to check qt5-requiring `libfive`? https://ports.macports.org/port/libfive/details

barracuda156 commented:
So it is not even `gnotime` itself that fails, but its archaic dependency `gnuregex`: https://trac.macports.org/ticket/62307

barracuda156 commented:
@mascguy When you have time, maybe you could help a bit here? I have set several ports to build with `guile2`, but several are left to be fixed, and some of those require Intel OS.

P. S. Few notes:

`gnotime` builds for me with the new `guile` (i.e. 3.0.9), but it is broken on Big Sur up: https://ports.macports.org/port/gnotime/details
For that reason, it failed when I added it to the PR, so I removed it for now.

For `autogen`, I could not force it to configure correctly yet.

mascguy commented:
> @mascguy When you have time, maybe you could help a bit here? I have set several ports to build with `guile2`, but several are left to be fixed, and some of those require Intel OS.
> 
> P. S. Few notes:
> 
> `gnotime` builds for me with the new `guile` (i.e. 3.0.9), but it is broken on Big Sur up: https://ports.macports.org/port/gnotime/details
> For that reason, it failed when I added it to the PR, so I removed it for now.
> 
> For `autogen`, I could not force it to configure correctly yet.

Sure, I'll try to find some time tomorrow after work.

mascguy commented on lang/guile2/Portfile:
```diff
@@ -0,0 +1,172 @@
+license             LGPL-2.1+
+
+version             2.2.7
+revision            3
```
Set revision to zero, since this is a new port

mascguy commented on lang/guile2/Portfile:
```diff
@@ -0,0 +1,172 @@
+homepage            https://www.gnu.org/software/guile/guile.html
+master_sites        gnu:guile
+
+depends_lib         port:readline \
```
Remember to use `-append` for all `depends_xxx`, `configure.args`, etc.

mascguy commented on lang/guile2/Portfile:
```diff
@@ -0,0 +1,172 @@
+    file delete ${destroot}${prefix}/share/info/guile.info-8
+    file delete ${destroot}${prefix}/share/info/guile.info-9
+    file delete ${destroot}${prefix}/share/info/guile.info-10
+    file delete ${destroot}${prefix}/share/info/r5rs.info
```
Rather than the copy-pasta, define all of these via a list, and then loop through them. Ditto for the file renames.

barracuda156 replied:
Checking if deleting works first. Have to rely on CI here, rebuilding `guile` takes about 10 hrs or more…

mascguy replied:
> Checking if deleting works first. Have to rely on CI here, rebuilding `guile` takes about 10 hrs or more…

One tip for these situations: Build a test stub port - with logic like this, that needs to be validated - so you don't have to depend on CI.

I'll post an example shortly.

mascguy replied:
Alternatively, an even easier approach is to test the logic directly in a TCL script. (While the various port-specific things like phases, etc, don't exist. But logic like looping, file deletes/renames, etc, all do. And you can simulate many things, by )

Here's an example I created back two years ago, when I was pre-testing logic for `openmpi`/`mpich`:

<details>
  <summary>mpi-testing.tcl</summary>
  
  ```tcl
#!/opt/local/bin/tclsh

if { ${argc} != 3 } {
    puts "Usage: ${argv0} <os.major> <os.arch> <debug-yes/no>"
    exit 1
}

# Define 'ui_debug', which is otherwise already available within a portfile
proc ui_debug {p_msg} {
    puts ${p_msg}
}

set os.major \
    [lindex $argv 0]
set os.arch \
    [lindex $argv 1]
set port_debug \
    [lindex $argv 2]

ui_debug "os_major: ${os.major}"
ui_debug "build_arch: ${os.arch}"
ui_debug "debug: ${port_debug}"

if { ![string is boolean -strict ${port_debug}] } {
    ui_debug "Error: arg 'debug' must be a boolean"
    exit 1
}

# Define 'variant_isset', which is otherwise already available within a portfile
proc variant_isset {p_debug} {
    global port_debug
    return ${port_debug}
}

proc mpiutil_add_subports {clist clist_unsupported clist_obsolete} {
    foreach key [dict keys ${clist}] {
        ui_debug "mpiutil_add_subport - clist - ${key}"
    }

    foreach key ${clist_unsupported} {
        ui_debug "mpiutil_add_subport - clist_unsupported - ${key}"
    }

    foreach key ${clist_obsolete} {
        ui_debug "mpiutil_add_subport - clist_obsolete - ${key}"
    }
}

set clist [dict create]
set clist_unsupported [list]
set clist_obsolete [list]

# TODO: Remove these obsolete subports on or after January 2022, along with all
# traces of -devel. For the latter, also remember to update proc
# 'mpiutil_add_subport', in portgroup 'mpiutil'.
lappend clist_obsolete \
    clang33 clang34 clang37 \
    clang50 clang60 clang70 clang80 \
    gcc49 gcc5 gcc6 gcc8

if { ${os.arch} eq "arm" } {
    # disable unreliable clang versions for ARM
    dict set clist gcc11   {macports-gcc-11}
    dict set clist clang11 {macports-clang-11}
    lappend clist_unsupported \
        clang90 clang10
} else {
    dict set clist clang90 {macports-clang-9.0}
    dict set clist clang10 {macports-clang-10}
    dict set clist clang11 {macports-clang-11}
}

# gcc 9+ only available on macOS10.7 (Darwin11) and newer
# However, gcc9+ builds fail on MacOS 10.7, so only enable for MacOS 10.8+.
if { ${os.arch} eq "arm" } {
    dict set clist gcc11 {macports-gcc-11}
    lappend clist_unsupported \
        gcc9 gcc10
} elseif { ${os.major} >= 12 } {
    dict set clist gcc9  {macports-gcc-9}
    dict set clist gcc10 {macports-gcc-10}
    dict set clist gcc11 {macports-gcc-11}
} else {
    lappend clist_unsupported \
        gcc9 gcc10 gcc11
}

ui_debug "clist: ${clist}"

mpiutil_add_subports \
    ${clist} ${clist_unsupported} ${clist_obsolete}
  ```
</details>

mascguy commented on lang/guile2/Portfile:
```diff
@@ -0,0 +1,172 @@
+master_sites        gnu:guile
+
+depends_lib         port:readline \
+                    port:gettext \
```
Unless there's something special about this port, `gettext` should be a build dep, while `gettext-runtime` should be a lib dep. (As the existing `guile` port is doing.)

mascguy commented on lang/guile2/Portfile:
```diff
@@ -0,0 +1,172 @@
+                    port:libunistring \
+                    port:boehmgc \
+                    port:libffi \
+                    port:ncurses
```
While this may seem nit-picky, it's preferable to alphabetize the various dependencies. Makes it easier to see what's what, particularly when the dependency list is lengthy.

mascguy commented on lang/guile/Portfile:
```diff
@@ -51,66 +26,92 @@ long_description    \
-    supported_archs ${build_arch}
-    set universal_archs_supported ${build_arch}
+                    --disable-silent-rules \
+                    --enable-shared
```
As with a dep list, alphabetize these as well. (Though `CPPFLAGS` and `LDDFLAGS` are fine to keep at the top of the list.)

mascguy commented on lang/guile/Portfile:
```diff
@@ -51,66 +26,92 @@ long_description    \
+    configure.args-append   --without-threads \
+                            --enable-jit=no \
+                            --disable-lto \
+                            --disable-nls
```
Alphabetize

mascguy commented on lang/guile2/Portfile:
```diff
@@ -0,0 +1,172 @@
+                    --mandir="${prefix}/share/man" \
+                    --enable-regex \
+                    --disable-error-on-warning \
+                    --disable-silent-rules
```
Alphabetize

mascguy commented on lang/guile/Portfile:
```diff
@@ -51,66 +26,92 @@ long_description    \
-build.env-append    SED=${prefix}/bin/gsed
+build.env-append     SED=${prefix}/bin/gsed
 
 configure.args      CPPFLAGS="-I${prefix}/include" \
```
Use `configure.args-append`

mascguy commented on lang/guile/Portfile:
```diff
@@ -51,66 +26,92 @@ long_description    \
+                    --enable-shared
+
+# configure may err on version mismatch otherwise
+patchfiles          patch-automake.diff
```
As with everything else, always use `patchfiles-append`.

barracuda156 commented:
@mascguy Thank you! I will make suggested corrections by tomorrow, today is pretty busy.

mascguy commented:
Reposting my TCL port testing example here, so it's not hidden once the various items are resolved.

<details>
  <summary>mpi-testing.tcl</summary>
  
  ```tcl
#!/opt/local/libexec/macports/bin/tclsh8.6

if { ${argc} != 3 } {
    puts "Usage: ${argv0} <os.major> <os.arch> <debug-yes/no>"
    exit 1
}

# Define 'ui_debug', which is otherwise already available within a portfile
proc ui_debug {p_msg} {
    puts ${p_msg}
}

set os.major \
    [lindex $argv 0]
set os.arch \
    [lindex $argv 1]
set port_debug \
    [lindex $argv 2]

ui_debug "os_major: ${os.major}"
ui_debug "build_arch: ${os.arch}"
ui_debug "debug: ${port_debug}"

if { ![string is boolean -strict ${port_debug}] } {
    ui_debug "Error: arg 'debug' must be a boolean"
    exit 1
}

# Define 'variant_isset', which is otherwise already available within a portfile
proc variant_isset {p_variant} {
    ui_debug "variant_isset: ${p_variant}"
    upvar ${p_variant} variant_var
    return ${variant_var}
}

set test_var [variant_isset port_debug]
ui_debug "Testing variant_isset port_debug: ${test_var}"

proc mpiutil_add_subports {clist clist_unsupported clist_obsolete} {
    foreach key [dict keys ${clist}] {
        ui_debug "mpiutil_add_subport - clist - ${key}"
    }

    foreach key ${clist_unsupported} {
        ui_debug "mpiutil_add_subport - clist_unsupported - ${key}"
    }

    foreach key ${clist_obsolete} {
        ui_debug "mpiutil_add_subport - clist_obsolete - ${key}"
    }
}

set clist [dict create]
set clist_unsupported [list]
set clist_obsolete [list]

# TODO: Remove these obsolete subports on or after January 2022, along with all
# traces of -devel. For the latter, also remember to update proc
# 'mpiutil_add_subport', in portgroup 'mpiutil'.
lappend clist_obsolete \
    clang33 clang34 clang37 \
    clang50 clang60 clang70 clang80 \
    gcc49 gcc5 gcc6 gcc8

if { ${os.arch} eq "arm" } {
    # disable unreliable clang versions for ARM
    dict set clist gcc11   {macports-gcc-11}
    dict set clist clang11 {macports-clang-11}
    lappend clist_unsupported \
        clang90 clang10
} else {
    dict set clist clang90 {macports-clang-9.0}
    dict set clist clang10 {macports-clang-10}
    dict set clist clang11 {macports-clang-11}
}

# gcc 9+ only available on macOS10.7 (Darwin11) and newer
# However, gcc9+ builds fail on MacOS 10.7, so only enable for MacOS 10.8+.
if { ${os.arch} eq "arm" } {
    dict set clist gcc11 {macports-gcc-11}
    lappend clist_unsupported \
        gcc9 gcc10
} elseif { ${os.major} >= 12 } {
    dict set clist gcc9  {macports-gcc-9}
    dict set clist gcc10 {macports-gcc-10}
    dict set clist gcc11 {macports-gcc-11}
} else {
    lappend clist_unsupported \
        gcc9 gcc10 gcc11
}

ui_debug "clist: ${clist}"

mpiutil_add_subports \
    ${clist} ${clist_unsupported} ${clist_obsolete}
  ```
</details>

mascguy commented:
Sergey, and progress on updating this port, per the suggestions?

barracuda156 commented:
> Sergey, and progress on updating this port, per the suggestions?

Allow me until this weekend, please :) 
This is a complicated task and I need a calm time to concentrate on the thing.
I want to complete that myself too, finally, so it won’t be left unattended.

Did you get a chance to see what can be done about `autogen` and a couple of ports that need Qt5 that depend on Guile?

mascguy commented:
> Allow me until this weekend, please :) This is a complicated task and I need a calm time to concentrate on the thing. I want to complete that myself too, finally, so it won’t be left unattended.
> 
> Did you get a chance to see what can be done about `autogen` and a couple of ports that need Qt5 that depend on Guile?

I re-reviewed the older comments, and didn't see anything related to your question. Can you refresh my memory, or provide a link to the discussion?

barracuda156 commented:
> I re-reviewed the older comments, and didn't see anything related to your question. Can you refresh my memory, or provide a link to the discussion?

With Guile updated, we need to either ensure dependencies build with Guile 3.0.9 (new `guile` port) or make them use `guile2`.
I have done the latter for several ports for which commits are here in the PR.

There are three related issues so far:
1. `gnotime` builds for me with a new Guile, but is apparently broken on newer MacOS, unrelated to Guile version used (see port health). So it won’t pass CI.
2. I tried to force `autogen` use `guile2` or adjust it to a new Guile, but so far could not get around its build system preferences. Will give it another go though.
3. There are 2–3 dependencies of Guile which cannot be built on PPC due to Qt5 requirement (as I recall, or something similar, will check when I am not on a mobile). So I cannot test them on my working setup and won’t be able to work on them. This is what I hope someone to help testing with.

mascguy commented:
> > I re-reviewed the older comments, and didn't see anything related to your question. Can you refresh my memory, or provide a link to the discussion?
> 
> With Guile updated, we need to either ensure dependencies build with Guile 3.0.9 (new `guile` port) or make them use `guile2`. I have done the latter for several ports for which commits are here in the PR.
> 
> There are three related issues so far:
> 
>     1. `gnotime` builds for me with a new Guile, but is apparently broken on newer MacOS, unrelated to Guile version used (see port health). So it won’t pass CI.
> 
>     2. I tried to force `autogen` use `guile2` or adjust it to a new Guile, but so far could not get around its build system preferences. Will give it another go though.
> 
>     3. There are 2–3 dependencies of Guile which cannot be built on PPC due to Qt5 requirement (as I recall, or something similar, will check when I am not on a mobile). So I cannot test them on my working setup and won’t be able to work on them. This is what I hope someone to help testing with.

Don't worry about the dependent ports yet, I can help with that. First just cleanup the `guile2` port, per my comments. Then we'll tackle the dependents.

barracuda156 commented:
@mascguy Looks like it works fine. Is post-destroot okay now?

mascguy commented:
> @mascguy Looks like it works fine. Is post-destroot okay now?

Going to do some local testing with all of this, now that the commits have been squashed down a bit. More to follow.

barracuda156 commented:
> > @mascguy Looks like it works fine. Is post-destroot okay now?
> 
> 
> Going to do some local testing with all of this, now that the commits have been squashed down a bit. More to follow.

Sounds good!

barracuda156 commented:
@mascguy Thank you for your help with this!

mascguy commented:
This has turned into a marathon mini-project, as there are a LOT of dependents to test. (And several are proving to be less-than-cooperative, in terms of providing a custom path to the appropriate version of Guile.) So I'm going to take a break from this, and revisit over the next day or two.

p.s. We shouldn't rush this anyway, to avoid breaking anything.

mascguy commented:
Sergey, I'm still committed to this PR though. So this is one less for you to worry about!

barracuda156 commented:
@mascguy Great. This one I cannot fully handle on my own: fixing Qt5 for PPC is not even on long-term todo list :)

barracuda156 commented:
@mascguy Please let me know if I can do something more for this.

mascguy commented:
> @mascguy Please let me know if I can do something more for this.

No, all set here. Will be returning to this soon.

kencu commented on lang/guile-3.0/Portfile:
```diff
@@ -0,0 +1,155 @@
+# configure may err on version mismatch otherwise
+patchfiles-append   patch-automake.diff
+
+if {${build_arch} in [list i386 ppc]} {
```
configure.build_arch

catap replied:
@kencu I'm wondering that is the differences between `configure.build_arch` and just `build_arch`?

mascguy replied:
> @kencu I'm wondering that is the differences between `configure.build_arch` and just `build_arch`?

@MarcusCalhoun-Lopez provided a great explanation, via his recent PR to fix these:

https://github.com/macports/macports-ports/pull/18857#issuecomment-1566212319

kencu commented on lang/guile-3.0/Portfile:
```diff
@@ -0,0 +1,155 @@
+}
+
+platform darwin {
+    if {${build_arch} eq "ppc"} {
```
configure.build_arch

mascguy replied:
> configure.build_arch

Fixed, thanks Ken!

barracuda156 commented:
@mascguy Where are we on this, Christopher? 

P. S. I can change `build_arch` to `configure.build_arch` any time, but do not want to rebase unnecessarily, it will trigger rebuilds.

mascguy commented:
> @mascguy Where are we on this, Christopher?

Given the potential for widespread breakage if rushed, I need time to thoroughly validate all of this locally. And simply haven't had the time yet.

Please be patient.  ;-)

> P. S. I can change `build_arch` to `configure.build_arch` any time, but do not want to rebase unnecessarily, it will trigger rebuilds.

Just fixed that, along with merge conflicts.

barracuda156 commented:
@mascguy Got it, thank you! No hurry, whenever you get time.

pmetzger commented:
@mascguy How's this looking?

pmetzger commented:
I think you may be taking on more than is humanly reasonable here; there are a lot of things you've put yourself into the critical path for.

mascguy commented:
> I think you may be taking on more than is humanly reasonable here; there are a lot of things you've put yourself into the critical path for.

I'm more than happy to step aside, if someone else would like to take on this responsibility. With the caveat that whoever merges this, needs to commit to immediately fixing any issues with the various dependents.

Would you like to step up to the plate Perry?  ;-)

pmetzger commented:
@mascguy No way, I don't have time either. My point is, if you commit to doing far more than any human can do, you end up with loads of dropped commitments across the project. I think you should try to figure out what is and isn't important and drop the commitments you can't keep.

mascguy commented:
Well, here's the thing: We have literally several THOUSAND open tickets. And based on my experience last year - where I random reviewed a few hundred of them - there are a large percentage that are no longer relevant. (Either because the issue was fixed without closing the ticket, or because it was related to an old version of the port.)

So perhaps you might want to tackle those instead, rather than ruffling everyone's feathers?

Just a thought...

pmetzger commented:
We have about 130 open PRs. The goal is to have about 30. This is humanly doable. I'm working on it now.

mascguy commented:
> We have about 130 open PRs. The goal is to have about 30. This is humanly doable. I'm working on it now.

Sure, and we all appreciate your efforts! But some patience and understanding would help, for more involved PRs like this one.

As for our massive ticket backlog though, it's something that's long-overdue for attention from all of us. Granted, ticket review can be a bit tedious and mind-numbing after a few hours. (Speaking from experience, after focusing on it for a few days last year.) Particularly if it requires reviewing the commit history for a given port.

But wow, if we could get the number of open tickets down to a manageable number... it would make a big difference!

I'm planning on bringing this up via the Dev mailing list soon. But figured I'd mention it, for anyone interested...

pmetzger commented:
Most of the tickets are years old and probably just should be closed at this point. Grooming the backlog would be very appreciated. FWIW, I think it would be very useful to move from Trac to Github for the tickets; it would make managing them much simpler. Much of the reason there are so many is, I think, the friction involved in managing the queue.

mascguy commented:
Finally have some free time, and will tackle this over the next few days.

barracuda156 commented:
@mascguy Awesome!

mascguy commented:
Currently working on this, and making progress. There are a few dependents that require closer scrutiny - such as `swig` and `swig3` - but so far, so good. More to follow.

mascguy commented:
Sergey, I'm seeing an issue with `guile-3.0`, which prevents successful use by dependents:

```
$ guile-config-3.0 compile
error: ("/opt/local/bin/pkg-config" "--cflags" "guile-3.0") exited with non-zero error code 127
```

Whereas `guile-2.2` works fine:

```
$ guile-config-2.2 compile
-D_THREAD_SAFE -I/opt/local/include/guile/2.2 -I/opt/local/include
```

Can you test locally, and see whether `guile-config-3.0 compile` works properly? (You don't need to build with the latest changes pushed here, though.)

barracuda156 commented:
@mascguy I shall, but I can do it on Oct 2 (i.e. once back to my hardware). I will make a notice to do it as soon as I return.

barracuda156 commented:
@mascguy I can confirm the same on `ppc`:
```
36-185% /opt/local/bin/guile-config compile
error: ("/opt/local/bin/pkg-config" "--cflags" "guile-3.0") exited with non-zero error code 127
36-185% /opt/local/bin/guile-config-2.2 compile
-D_THREAD_SAFE -I/opt/local/include/guile/2.2 -I/opt/local/include
```

barracuda156 commented:
@mascguy It should work in fact: explicit invocation of the command in question works correctly:
```
36-185% pkg-config --modversion guile-3.0
3.0.9
36-185% pkg-config --cflags guile-3.0
-I/opt/local/include/guile/3.0 -I/opt/local -I/opt/local/include
```
And `guile-config` is just a text script, not an executable. Not sure where it goes wrong.

barracuda156 commented:
And also, it explicitly states in itself: `This script has been deprecated. Just use pkg-config`.
Perhaps we do not even need to bother about it.

dglb commented:
> @barracuda156 does any port depends on `guile`? And if so, can it be built with your update?

the package 
 mu +guile
depends on guile 3.0

in log
 :info:configure Run-time dependency guile-3.0 found: NO (tried pkgconfig, framework and cmake)

It's a upstream dependency...
https://github.com/djcb/mu
 mu-guile: bindings for the Guile/Scheme programming language (version 3.0 and later)

thanks,

barracuda156 commented:
@dglb Did you try this PR, which does introduce guile 3.0.x?

barracuda156 commented:
@mascguy Should I rebase to resolve conflicts, or too early anyway?

mascguy commented:
> @mascguy Should I rebase to resolve conflicts, or too early anyway?

Don't worry about it, I'll take care of it when I have time

dglb commented:
> @dglb Did you try this PR, which does introduce guile 3.0.x?

I did it and it's working: I can compile mu +guile.
  mu @1.10.7_0+emacs+guile (active)

thank's

I hope it will be soon in the "real" macports tree.

pmetzger commented:
This is closing in on two years old. I would like to suggest that we close it for now, and that when someone has time to work on the problem, they can reopen it.

barracuda156 commented:
@pmetzger We do not gain anything by having all this wasted. In the worst case guile3 can be added as a new port, which will not require any changes to existing ones.

pmetzger commented:
@barracuda156 Nothing is wasted. It's simply a change of notation on the pull request. Nothing prevents continuing work, changing the state back to open, opening new PRs, etc.

I do want to emphasize that the PR queue is meant for things that can be pulled now. The proliferation of drafts that are never progressed is not really what the queue is for. If one is working on a project, one can do so at will without having to invoke the queue. One can even run the CI in one's own fork of the repo at will.

However on the objeject level here, given the length of time involved, I think your approach of creating a guile3 port and then moving things over one by one seems least intrusive.

pmetzger commented:
As explained above, I'm closing this for now. It can be re-opened when someone wants to, or a guile3 port can be started to make the transition easier, or some other method can be chosen.

lemzwerg commented:
It's great that someone is working on the transition to guile 3, thanks for that!

The `configure` script of LilyPond's development version no longer checks for guile-2; you have to explicitly enforce it.  I thus wonder whether there is any progress with this patch.

barracuda156 commented:
@lemzwerg There are two options here:

1. Someone steps in to help with moving ports to guile 3.x, whatever possible, the rest are switched to use guile2. This is what has been partially done here, but it got stuck for months, which was the reason the PR got closed.
I cannot handle everything myself and test on every existing system. Presumably maintainers of the ports relying on guile should be interested in moving to the current version, but nobody aside of @mascguy and myself joined in.

A technical complication is that some number of guile-related ports are likely to be just broken (unrelated to this PR), and that will prevent CI from passing. I have no write access to the repo, so cannot help here. Someone should make a judgement whether specific ports are fixable, and if not, merge the PR regardless.

2. Instead, guile3 can be made as a new port without switching anything to using it (at least immediately). This will not require fixing archaic broken ports or testing anything besides guile3 itself. This can be easily accomplished – we have everything for that already here – but we need some consensus on the matter.
@mascguy @pmetzger @catap Any opinions?

lemzwerg commented:
Alas, I'm not a real Mac user at all, just helping maintain the LilyPond port; I have only sporadic access to an old Lion box, which is not really helpful here.

BTW, Perry already voiced his opinion [above](https://github.com/macports/macports-ports/pull/14600#issuecomment-1871290897) to proceed with your option 2.

lemzwerg commented on textproc/lilypond/Portfile:
```diff
@@ -106,11 +106,12 @@ set py_ver_nodot    [string map {. {}} ${py_ver}]
 
 # Either 'pango' or 'pango-devel' will do, thus we depend on its
 # pkgconfig file.
+set guile_ver       3.0
```
I think it's better if you restrict guile 3 to the `lilypond-devel` subport.

lemzwerg commented:
Ping!

barracuda156 commented:
> BTW, Perry already voiced his opinion [above](https://github.com/macports/macports-ports/pull/14600#issuecomment-1871290897) to proceed with your option 2.

@lemzwerg I will try to get this sorted again, ok. In a course of a few days – I am busy with other stuff atm.

mascguy commented:
> Sergey, I'm seeing an issue with `guile-3.0`, which prevents successful use by dependents:
> 
> ```
> $ guile-config-3.0 compile
> error: ("/opt/local/bin/pkg-config" "--cflags" "guile-3.0") exited with non-zero error code 127
> ```
> 
> Whereas `guile-2.2` works fine:
> 
> ```
> $ guile-config-2.2 compile
> -D_THREAD_SAFE -I/opt/local/include/guile/2.2 -I/opt/local/include
> ```

Just a friendly reminder: What ultimately stopped progress on this, is issues with Guile 3.x. (And back when I investigated 6+ months ago, it also appeared to affect the Guile 3.0 cask from brew. So it's not a MacPorts-specific issue.)

Haven't returned to it since, so there's a chance it's been fixed by upstream. But if not, then we need to engage with upstream, to get it fixed.

barracuda156 commented:
@mascguy I think that at least on our side chances of it getting fixed increase substantially with having a port to begin with. This of course does not mean that other ports should be switched to the new Guile without verifying if that works, as it happens with some other languages )
But introducing it in a non-destructive way should be beneficial.

mascguy commented:
> > @mascguy I think that at least on our side chances of it getting fixed increase substantially with having a port to begin with. This of course does not mean that other ports should be switched to the new Guile without verifying if that works, as it happens with some other languages ) But introducing it in a non-destructive way should be beneficial.
> 
> Well, the problem is that it's pointless to add a Guile 3.x port, until/unless it actually works. And the failure mentioned earlier is fatal, blocking any potential use by dependents.

Although you're right, it's a chicken-and-egg issue. So we can certainly add a Guile 3.x port in isolation, so long as the port clearly states that it's not ready for consumption.

How does that sound?

barracuda156 commented:
@mascguy Yes, as the first step it will be good, and done that way it is safe.

barracuda156 commented:
> Another ping – guile is now available as version 3.0.10...

I need to build it locally first. It takes _a lot_ of time, but I will try to make it in coming days.

pmetzger commented:
@lemzwerg You might try making a pull request of your own.

lemzwerg commented:
@pmetzger Sorry, but I have no access to a Mac right now, and I'm actually not a Mac user at all – I just try to take care of LilyPond, which needs guile 3 for current development versions.

barracuda156 commented:
@lemzwerg I will try to make it in coming days. As discussed above, it will not be in a form of this (closed) PR, but rather as a new `guile3` port.

barracuda156 commented:
@graywolf Oh, great, please handle this.

It would be great if you pick fixes for legacy systems from my portfile (though if it is problematic, I can do a separate PR for those).

barracuda156 commented:
@graywolf

> What is the repository (powerpc-ports)?

`powerpc-ports` is just a name I use for the fork because my focus is on powerpc. You can fork a repo, but change its name. Name and contents of other branches in my repo is irrelevant, this PR was done against MacPorts master.

> Since you apparently already did the work (I see guile-2.2), why did you not sent it to this repository?

Well, as you can see, the PR here was not merged. I kept the upgrade locally.

> Would you know why dprintf is misdetected? Would you happen to know why Oresolve-primitives -Ocps is required on i386 and ppc? Same with the -O1.

If the flags are present in earlier guile ports, they are just borrowed. If not, then they were added to fix some errors. Since many month passed, now I cannot say why something was done. For sure nothing was added just for the sake of making a port more complicated.
Unless there are strong objections regarding specific flags, I would retain what I had, since these are confirmed to produce a working build. `guile` takes _many_ hours to build on PowerPC, since it has to make a full bootstrap. I am not ready to spend several days of compilation time to check if alternative combinations of flags gonna work or not.

barracuda156 commented:
Yeah, if it was not explicit, I am not expecting or asking anyone to test this for ppc or introduce specific fixes.
I merely meant to copy-paste what was already done. Nothing beyond that.
